Mesh component

Tjalling-dejong opened the pull request:
## Issue addressed
Adresses issue #671 
Implementation of the ModelComponent class for the mesh component.

## Explanation
Explain how you addressed the bug/feature request, what choices you made and why.

## Checklist
- [x] Updated tests or added new tests
- [ ] Branch is up to date with `main`
- [x] Tests & pre-commit hooks pass
- [ ] Updated documentation if needed
- [ ] Updated changelog.rst if needed
- [ ] For predefined catalogs: update the catalog version in the file itself, the references in data/predefined_catalogs.yml, and the changelog in data/changelog.rst

## Additional Notes (optional)
Add any additional notes or information that may be helpful.

Tjalling-dejong commented:
@hboisgon I am encountering an issue with the get_mesh() method, I saw through git blame that you made the last commit to this code. I have not changed the code. Now that we are using XUgrid version 0.0.9 the code breaks. I did not manage to figure it out if the bug is from our side or in XUgrid. Could you maybe have a look?

hboisgon commented:
> @hboisgon I am encountering an issue with the get_mesh() method, I saw through git blame that you made the last commit to this code. I have not changed the code. Now that we are using XUgrid version 0.0.9 the code breaks. I did not manage to figure it out if the bug is from our side or in XUgrid. Could you maybe have a look?

I see we get the same error in the hydromt_delft3dfm test. I'll have a look but in main and not v1

deltamarnix commented on hydromt/models/__init__.py:
```diff
@@ -8,9 +8,8 @@
-if HAS_XUGRID:
-    from hydromt.models.components.mesh import MeshModel
-
+# if HAS_XUGRID:
+#     from hydromt.models.components.mesh import MeshModel
```
Remove?

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
 
-    ## general setup methods
-    def setup_mesh2d_from_rasterdataset(
+    def set(
+        self,
+        data: Union[xu.UgridDataArray, xu.UgridDataset],
+        name: Optional[str] = None,
+        grid_name: Optional[str] = None,
+        overwrite_grid: bool = False,
+    ) -> None:
```
```suggestion
    def set(
        self,
        data: Union[xu.UgridDataArray, xu.UgridDataset],
        *,
        name: Optional[str] = None,
        grid_name: Optional[str] = None,
        overwrite_grid: bool = False,
    ) -> None:
```

I would make all naming mandatory

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+            for grid in self.data.ugrid.grids:
+                grid.set_crs(crs)
+
+        # # update related geoms if necessary: region TODO: Check if is still needed
+        # if overwrite_grid or new_grid:
+        #     # add / updates region
+        #     if "region" in self.geoms:
+        #         self._geoms.pop("region", None)
+        #     _ = self.region
```
Is it still needed? ;)

hboisgon replied:
No but then I suggested to add a new ``mesh_region`` property to replace the old ``region`` one. It will be quite useful to update ``MeshComponentRegion`` as we discussed.

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        data = _check_UGrid(data, name)
+
+        # Checks on grid topology
+        # TODO: check if we support setting multiple grids at once. For now just one
+        if len(data.ugrid.grids) > 1:
+            raise ValueError(
+                "set_mesh methods only supports adding data to one grid at a time."
+            )
```
What would that entail? I can imagine that we want one grid per component.

Tjalling-dejong replied:
I don't know. The TODO was already there in the original code and left it in for now.

hboisgon replied:
For now ``MeshComponent`` is still multi-grid. We can see later if we want to add single grid Component like ``Mesh1dComponent``, ``Mesh2dComponent`` and ``Mesh3dComponent``. The TODO can be removed and open as an issue if anyone complains. It's just that for the set methods I only allow to update data or grid schematization one grid at a time.

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        elif grid_name != data.ugrid.grid.name:
+            data = rename_mesh(data, name=grid_name)
+
+        # new_grid = grid_name not in self.mesh_names
```
Remove?

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        overwrite_grid: bool, optional
+            If True, overwrite the grid with the same name as the grid in self.mesh.
+        """
+        # Checks on data
```
I expected the first call to be `_initialize_mesh()`, just like other components.

Tjalling-dejong replied:
I think it is done this way because you cannot create an empty UgridDataset

hboisgon replied:
Indeed! So the choice would either be xr.Dataset or None but I think None is easier and clearer to handle.

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        if self.data is None:
+            self.logger.debug("No mesh data found, skip writing.")
+            return
+        if not self.model_root.is_writing_mode():
+            raise IOError("Model opened in read-only mode")
```
```suggestion
    self._model_root._assert_write_mode()
```

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        **kwargs : dict
+            Additional keyword arguments to be passed to the `read_nc` method.
+        """
+        if not self.model_root.is_reading_mode():
+            raise IOError("Model not opend in read mode")
```
```suggestion
    self._model_root._assert_read_mode()
```

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+            raise IOError("Model not opend in read mode")
+        ds = xr.merge(
+            read_nc(
+                fn, root=self.model_root.path, logger=self.logger, **kwargs
```
```suggestion
                fn, root=self._model_root.path, logger=self._logger, **kwargs
```

Do you have tests for these, or should you update from the refactor-models branch?

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+                )
+        self._data = uds
+
+    def create(
```
In my branch `region-component-2`, I am trying to get rid of the `region: dict`. I think this will influence this piece of code as well.

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+            return self.data.ugrid.bounds
+        return None
+
+    @property
+    def region(self) -> gpd.GeoDataFrame:
+        """Returns geometry of region of the model area of interest based on mesh total bounds."""  # noqa: E501
+        region = gpd.GeoDataFrame()
+        if self.data is not None:
+            region = gpd.GeoDataFrame(
+                geometry=[box(*self.data.ugrid.total_bounds)], crs=self.crs
+            )
+        return region
```
Shouldn't this be handled by the `region` component instead?

deltamarnix commented on hydromt/models/components/mesh.py:
```diff
@@ -10,34 +10,389 @@
+        else:
+            return self.mesh_grids[grid_name]
+
+    def add_data_from_rasterdataset(
         self,
         raster_fn: Union[str, Path, xr.DataArray, xr.Dataset],
         grid_name: Optional[str] = "mesh2d",
         variables: Optional[list] = None,
         fill_method: Optional[str] = None,
```
```suggestion
    def add_data_from_rasterdataset(
        self,
        *,
        raster_fn: Union[str, Path, xr.DataArray, xr.Dataset],
        grid_name: Optional[str] = "mesh2d",
        variables: Optional[list] = None,
        fill_method: Optional[str] = None,
```

deltamarnix commented on tests/conftest.py:
```diff
@@ -363,3 +367,15 @@ def artifact_data():
+
+
+@pytest.fixture()
+def mock_model(tmpdir):
```
There are already some other mocks for a Model. Not sure if you want to put this in conftest, as it will be different per test case.

Tjalling-dejong replied:
I'm using this fixture in test_grid_component as well. Should I use I put this fixture in both test_mesh_component.py and test_grid_component.py?

deltamarnix commented on tests/conftest.py:
```diff
@@ -363,3 +367,15 @@ def artifact_data():
+    logger = logging.getLogger(__name__)
+    logger.propagate = True
+    model = create_autospec(Model)
+    model.root = ModelRoot(path=tmpdir)
+    model.data_catalog = DataCatalog()
+    model.region = ModelRegionComponent(model=model)
```
Let's mock the other parts as well

deltamarnix commented on tests/conftest.py:
```diff
@@ -363,3 +367,15 @@ def artifact_data():
+
+
+@pytest.fixture()
+def mock_model(tmpdir):
```
Try using `def mock_model(tmpdir, mocker: MockerFixture):`, so that you can use the `pytest-mock` wrapper for making your mocks.

deltamarnix reviewed (commented):
This is my initial review. I will probably come with more in a second round, but that also depends on my own experiences with creating components.

Let's have another look at how we mock in our repository. I prefer using `pytest-mock`, where we make use of `MockerFixture`. In that case we don't need to write things like: `with patch(...):`.

savente93 commented on hydromt/models/plugins.py:
```diff
@@ -13,7 +13,6 @@
 # local generic models
 LOCAL_EPS = {
     "vector_model": "hydromt.models.components.vector:VectorModel",
-    "mesh_model": "hydromt.models.components.mesh:MeshModel",
     "network_model": "hydromt.models.components.network:NetworkModel",
 }
```
I can see this code is not up to date with `model-refactor`. if we're going to wait to get that in before you merge this you might want to merge it because I can already see it's going to cause a few merge issues here.

savente93 commented on hydromt/models/components/mesh.py:
```diff
@@ -199,113 +554,41 @@ def setup_mesh2d_from_raster_reclass(
-                ):
-                    if not overwrite_grid:
+                if not self._grid_is_equal(grid_name, data):
+                    if not overwrite_grid:  # TODO: add is_overwrite_mode
```
Fix todo, or make issue?

hboisgon replied:
Why was the is_overwrite_mode added? Calling the whole ``set`` method means you are overwriting anyway if not the grid then the data...

Tjalling-dejong replied:
I removed is_overwrite_mode. Not sure why I added that

savente93 reviewed (commented):
Nice tests. I have a few small comments but once it's up to date it should be very close.

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+    rename_mesh,
+)
 
-__all__ = ["MeshModel"]
```
I think you still need the __all__ but his time with MeshModelComponent.
@savente93 previously xugrid was an optional package. How does this work now with components discovery? Can we still make this one discovered only if xugrid is installed? Or should in the end xugrid become a mandatory dependency of hydromt ?

Tjalling-dejong replied:
XUgrid will become a mandatory dependency in v1

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+        elif grid_name != data.ugrid.grid.name:
+            data = rename_mesh(data, name=grid_name)
+
+        crs = self._add_mesh(
+            data=data, grid_name=grid_name, overwrite_grid=overwrite_grid
+        )
+
+        if crs:  # Restore crs
+            for grid in self.data.ugrid.grids:
+                grid.set_crs(crs)
```
Nice that you did a new function, I would however also include updating crs in that function as well instead of here.

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+            for grid in self.data.ugrid.grids:
+                grid.set_crs(crs)
+
+    def write(
```
You'll need to add ``@hydromt_step`` decorator to all methods available for build and update ie read/write/create/add_data

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+                )
+        self._data = uds
+
+    def create(
```
This is to create a 2D mesh so I would keep it in the function name. the current MeshComponent handles multiple grids which can be 1D or 2D or 3D.

```suggestion
    def create2d(
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+        else:
+            return self.mesh_grids[grid_name]
+
+    def add_data_from_rasterdataset(
```
```suggestion
    @hydromt_step
    def add_2ddata_from_rasterdataset(
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+
+    def add_data_from_rasterdataset(
+        self,
+        *,
```
Maybe the * can go after raster_fn? We're already not happy with the name :)

deltamarnix replied:
It almost looks like these are two different functions. One for a file, one for an actual dataset

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
         variables: Optional[list] = None,
         fill_method: Optional[str] = None,
-        resampling_method: Optional[Union[str, List]] = "centroid",
+        resampling_method: Optional[str] = "centroid",
```
It can still be a list right? In case method is different per variable in ``variables``

Tjalling-dejong replied:
I based the typing of the typing in mesh2d_from_rasterdataset. There it is Optional[str], but I see in the docstring now that a list is also possible. I'll correct it in mesh2d_from_rasterdataset as well

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -78,35 +415,35 @@ def setup_mesh2d_from_rasterdataset(
         return list(uds_sample.data_vars.keys())
 
-    def setup_mesh2d_from_raster_reclass(
+    def add_data_from_raster_reclass(
```
```suggestion
    @hydromt_step
    def add_2ddata_from_raster_reclass(
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -195,117 +532,45 @@ def setup_mesh2d_from_raster_reclass(
             # Check on crs
             if not data.ugrid.grid.crs == self.crs:
-                raise ValueError("Data and self.mesh should have the same CRS.")
+                raise ValueError("Data and self.data should have the same CRS.")
```
```suggestion
                raise ValueError("Data and Mesh should have the same CRS.")
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -11,33 +11,370 @@
+        if self.data is not None:
+            return self.data.ugrid.bounds
+        return None
+
```
Since you removed ``region`` maybe you could create an equivalent ``mesh_region`` property? 
```suggestion
    @property 
    def mesh_region(self) -> Optional[gpd.GeoDataFrame]:
        """Return mesh total_bounds as a geodataframe."""
        if self.data is not None:
            region = gpd.GeoDataFrame(
                geometry=[box(*self.data.ugrid.total_bounds)], crs=self.crs
            )
            return region
        return None
```

hboisgon reviewed (changes requested):
Nice work @Tjalling-dejong especially with the extra testing!
A few comments to address before merging but the refactor PR changed a little again so I think you'll have to wait for this one to be merged first before you can address everything.

deltamarnix commented on hydromt/components/mesh.py:
```diff
@@ -12,33 +12,376 @@
 
-__all__ = ["MeshModel"]
-logger = logging.getLogger(__name__)
+logger = getLogger(__name__)
```
I believe that this `logger` can go

deltamarnix commented on hydromt/components/mesh.py:
```diff
@@ -12,33 +12,376 @@
+
+    def write(
+        self,
+        fn: str = "mesh/mesh.nc",
```
I would opt for a `DEFAULT_FN = "mesh/mesh.nc` at the top of this file and use it in `fn` for both the `read()` and `write()` functions.

deltamarnix commented on hydromt/components/mesh.py:
```diff
@@ -12,33 +12,376 @@
+        method.
+        """
+        # XU grid data type Xarray dataset with xu sampling.
+        if self._data is None and self._root.is_reading_mode():
+            self.read()
```
I miss this `_initialize_mesh` function that we also see in the grid component. Would be nice to have it consistent.

deltamarnix commented on hydromt/components/mesh.py:
```diff
@@ -12,33 +12,376 @@
+                if v == crs:
+                    continue
+                else:
+                    raise ValueError(
+                        f"Mesh crs is not uniform, please check {grid_crs}"
+                    )
```
I would expect this in a setter, not in a getter.

Tjalling-dejong replied:
I think this feature can go. There is a check in _add_mesh that checks if the crs is uniform.

deltamarnix commented on hydromt/components/mesh.py:
```diff
@@ -12,33 +12,376 @@
+        else:
+            return None
+
+    @property
+    def bounds(self) -> Optional[Dict]:
+        """Returns model mesh bounds."""
+        if self.data is not None:
+            return self.data.ugrid.bounds
+        return None
+    @property 
+    def mesh_region(self) -> Optional[gpd.GeoDataFrame]:
+        """Return mesh total_bounds as a geodataframe."""
+        if self.data is not None:
+            region = gpd.GeoDataFrame(
+                geometry=[box(*self.data.ugrid.total_bounds)], crs=self.crs
+            )
+            return region
+        return None
+    @property
+    def mesh_names(self) -> List[str]:
+        """List of grid names in mesh."""
+        if self.data is not None:
+            return [grid.name for grid in self.data.ugrid.grids]
+        else:
+            return []
```
Would be nice if your editor did autoformatting. We have enabled that in vs code settings. Perhaps you have some overwritten settings in your vs code personal settings. I expect some new lines in between

Tjalling-dejong replied:
Weird that it did not autoformat this time

Tjalling-dejong commented:
@deltamarnix @hboisgon @savente93  Thanks for the feedback!
I added your suggestions and updated code + tests. 
There are still multiple breaking tests but these are related to plugins and circular imports. I will look into that tomorrow.

hboisgon commented on docs/dev/migrating-to-v1.rst:
```diff
@@ -240,6 +240,7 @@ In the core of HydroMT, the available components are:
 | model.tables          | TablesComponent  | Component for managing non-geospatial data in pandas DataFrames  |
 | model.grid            | GridComponent   | Component for managing regular gridded data in single hydromt RasterDataset    |
 | model.geoms('region') | RegionComponent | Component for managing the area of interest for the model in a geopandas GeoDataFrame.    |
+| model.mesh			| MeshComponent	  | Component for managing unstructured grids as a hydromt RasterDataset
```
Table format looks off. Can you check? All columns length should match ie "|" should be aligned. If it was because of my suggestion then it's because I cant easily check alignment in this github view sorry..

hboisgon commented on docs/dev/migrating-to-v1.rst:
```diff
@@ -355,22 +356,36 @@ has not been changed compared to the GridModel.
+| model.set_grid(...)          | model.grid.set(...)             |
 +------------------------------+-------------------------------------------+
-| model.read_grid(...)         | model.grid_component.read(...)            |
+| model.read_grid(...)         | model.grid.read(...)            |
 +------------------------------+-------------------------------------------+
-| model.write_grid(...)        | model.grid_component.write(...)           |
+| model.write_grid(...)        | model.grid.write(...)           |
 +------------------------------+-------------------------------------------+
-| model.setup_grid(...)        | model.grid_component.create(...)          |
+| model.setup_grid(...)        | model.grid.create(...)          |
 +------------------------------+-------------------------------------------+
-| model.setup_grid_from_*(...) | model.grid_component.add_data_from_*(...) |
+| model.setup_grid_from_*(...) | model.grid.add_data_from_*(...) |
```
Same here for table alignement

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+class MeshComponent(ModelComponent):
+    """ModelComponent class for mesh components."""
+
+    def __init__(self, model: "Model"):
+        super().__init__(model=model)
```
We just discussed yesterday that we want to pass a default filename in the component __init__ and then use None in the argument for read and write. Can you update?
You can check out the TablesComponent PR

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+        """
+        # XU grid data type Xarray dataset with xu sampling.
+        if self._data is None:
+            self._initialize_mesh()
```
```suggestion
            self._initialize()
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+    def _initialize_mesh(self, skip_read: bool = False) -> None:
+        """Initialize mesh object."""
+        if self._data is None:
+            self._data = xu.UgridDataset(xr.Dataset())
```
Does this work now? Then nice we can initialize even with an empty grid :)
You'll need to add an initialize step in the read function. You can check maybe TablesComponent to see how initialize is used

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+    @property
+    def crs(self) -> Optional[CRS]:
+        """Returns model mesh crs."""
+        if self.data is not None:
```
Maybe you need to update all the properties now no? Because with initialize you will get an empty xugrid object and not None anymore for data. Check this also works for empty xugrid object.
Please update all properties like bounds etc accordingly

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+        return None
+
+    @property
+    def mesh_region(self) -> Optional[gpd.GeoDataFrame]:
```
I think this will be named region again in @deltamarnix PR on regions so maybe you could change the name already or check with him how to best update?

hboisgon commented on docs/api.rst:
```diff
@@ -553,6 +553,59 @@ General methods
+
+   MeshComponent.set
+   MeshComponent.write
+   MeshComponent.read
+
```
```suggestion
   MeshComponent.read
   MeshComponent.get_mesh
```

hboisgon commented on hydromt/components/mesh.py:
```diff
@@ -1,40 +1,384 @@
+            return self.mesh_grids[grid_name]
+
+    @hydromt_step
+    def add_data_from_rasterdataset(
```
I still miss in the name that this method works for 2D grids and not 1D and 3D. I previously suggested ``add_2ddata_from_rasterdataset``. I'm never good with names but I still think we should highlight this fact somehow in the name.

Tjalling-dejong replied:
I renamed it to `add_2d_data_from_rasterdataset`

hboisgon reviewed (changes requested):
Almost there @Tjalling-dejong ! Nice that you could add an initialize function for mesh! But I think you still need to update a couple of things to fully take it into account.
I wonder if we are not missing model update tests in our tests benches? Ie model opened in r+ mode and check if the existing mesh/grid/other components are indeed read properly before being updated. If these tests are in place please ignore.

hboisgon commented on tests/components/test_mesh_component.py:
```diff
@@ -166,6 +180,33 @@ def test_read(mock_model, caplog, tmpdir, griduda):
+    assert "elevation_mesh" in component.mesh_names
+    assert component.data == data
+    component.write()
+    component._data = None
+    component.read()
```
```suggestion
    component._data = None
    component.read()
    assert component.data == data
    
    m2 = Model(root=str(tmpdir), mode="r+")
    # Sample data
    data = xu.data.elevation_nl().to_dataset()
    data.grid.crs = 28992
    data.name = "elevation_v2"
    m.mesh.set(data=data, grid_name="elevation_mesh")
    
    # check both data variables are still there ie mesh was read before being updated
    assert "elevation_v2" in m2.mesh.data
    assert "elevation" in m2.mesh.data # I think this is the correct name?
    
```

hboisgon commented on docs/dev/migrating-to-v1.rst:
```diff
@@ -362,19 +367,20 @@ has not been changed compared to the GridModel.
 +------------------------------+-------------------------------------------+
-| model.setup_grid_from_*(...) | model.grid.add_data_from_*(...) |
+| model.setup_grid_from_*(...) | model.grid.add_data_from_*(...) 		   |
 +------------------------------+-------------------------------------------+
```
Alignment still looks off? Or is it just this view?

Tjalling-dejong replied:
It is just the view, if you check this file in the branch the table looks fine

hboisgon reviewed (changes requested):
Thanks for the changes @Tjalling-dejong. Almost there but still a couple of steps to add to fully take advantage of the new ``_initialize`` function.
I also explained the test I wanted in a little more details, sorry it was not clear enough.

If you update the initialize part then we are good to merge :)

hboisgon reviewed (approved):
Looking all good now, thanks @Tjalling-dejong :)
